app_test.go

```diff
@@ -473,3 +491,76 @@ func TestNopLogger(t *testing.T) {
 	app := fxtest.New(t, NopLogger)
 	app.RequireStart().RequireStop()
 }
+
+type testErrorWithGraph struct {
+	graph string
+}
+
+func (we testErrorWithGraph) Graph() DotGraph {
+	return DotGraph(we.graph)
+}
+
+func (we testErrorWithGraph) Error() string {
+	return "great sadness"
+}
+
+func TestVisualizeError(t *testing.T) {
+	t.Run("NotWrappedError", func(t *testing.T) {
+		graph, err := VisualizeError(errors.New("great sadness"))
+		assert.Empty(t, graph)
+		require.Error(t, err)
+	})
+
+	t.Run("WrappedErrorWithEmptyGraph", func(t *testing.T) {
+		graph, err := VisualizeError(testErrorWithGraph{graph: ""})
+		assert.Empty(t, graph)
+		require.Error(t, err)
+	})
+
+	t.Run("WrappedError", func(t *testing.T) {
+		graph, err := VisualizeError(testErrorWithGraph{graph: "graph"})
+		assert.Equal(t, "graph", graph)
+		require.NoError(t, err)
+	})
+}
+
+func TestErrorHook(t *testing.T) {
+	t.Run("UnvisualizableError", func(t *testing.T) {
+		type A struct{}
+
+		var errStr, graphStr string
+		h := errHandlerFunc(func(err error) {
+			errStr = err.Error()
+			graphStr, _ = VisualizeError(err)
+		})
+		fxtest.New(t,
+			Provide(func() A { return A{} }),
+			Invoke(func(A) error { return fmt.Errorf("great sadness") }),
+			ErrorHook(h),
+		)
+		assert.Equal(t, "great sadness", errStr)
+		assert.Empty(t, graphStr)
```

Comment:
In this case we should get the original error back (after you make the change
suggested above) so we can just do `assert.Equal` on the original error object.